add HAVING filtering to groupNode render

dt opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3547)

<!-- Reviewable:end -->

dt committed:
add HAVING filtering to groupNode render

tamird commented on sql/group.go:
```diff
@@ -195,6 +214,17 @@ func (n *groupNode) computeAggregates() {
+				n.err = err
+				return
+			}
+			if res == parser.DNull || !bool(res.(parser.DBool)) {
```
Consider exporting parser.getBool and using it here. It should be an error for `res` to not be a boolean, right? Add a test case to that effect as well.

tamird commented:
LGTM

dt committed:
Export parser.getBool

dt commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---

*[sql/group.go, line 223 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6j8Ohoad7T140WyMm4-r1-223:-K6jDIC8iOP4yEJd68Ev:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ed648f4c2d5d77fb0783c926f35d24fb7c27243/sql/group.go#L223)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r1, 5 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/group.go, line 223 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jE7E9ROu2HY4AJ80J:-K6jE7E9ROu2HY4AJ80K:-1058102021) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b639483d7505f172ed4d4a308f22f1a66bbecc2/sql/group.go#L223)):*
do you think it makes sense to put the NULL handling into `res`?

---

*[sql/parser/datum.go, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jDynjTE2t-X936cZo:-K6jDynjTE2t-X936cZp:-168154467) ([raw file](https://github.com/cockroachdb/cockroach/blob/4368db53d8ed19d146ad28a966c05622c8f448ee/sql/parser/datum.go#L87)):*
needs a comment to pass golint

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/group.go, line 223 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jE7E9ROu2HY4AJ80J:-K6jEMjVmY8gSeMiC1zK:847576884) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b639483d7505f172ed4d4a308f22f1a66bbecc2/sql/group.go#L223)):*
err, into GetBool

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 3 of 5 files reviewed at latest revision, all discussions resolved.

---

*[sql/group.go, line 223 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jE7E9ROu2HY4AJ80J:-K6jGx8bQm_aTirvAszt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b639483d7505f172ed4d4a308f22f1a66bbecc2/sql/group.go#L223)):*
Done.

---

*[sql/parser/datum.go, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jDynjTE2t-X936cZo:-K6jGzVk2_eQdVuGyjRq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4368db53d8ed19d146ad28a966c05622c8f448ee/sql/parser/datum.go#L87)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6jJhNV6Pe-wJLlU74t:-550821556) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
I believe `HAVING` acts like `WHERE` if there are no aggregate functions or group by clause.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6jQ15oJVPehW2vBPGM:485781513) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
This was supposed to test the boolean type requirement of HAVING expressions, eg with postgres `SELECT MAX(k), MIN(v) FROM kv HAVING k;` yields: 
```
ERROR:  argument of HAVING must be type boolean, not type integer
```

As far as `HAVING` with no GROUP or agg, looks like postgres errors, e.g. `SELECT k FROM kv HAVING k > 3;`:
```
ERROR:  column "kv.k" must appear in the GROUP BY clause or be used in an aggregate function
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6jRCrsRY9HVtYY_6mo:2105639360) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
Hmm, perhaps my memory is failing me. Regardless, there should be a test for this (error) case.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6js_kjJVrq7kkNtbmu:1114545475) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
After playing with postgres a bit: looks like a `HAVING` clause, even just `HAVING TRUE`, forces on aggregation (e.g. causes `SELECT k` to start requiring `k` appear in `GROUP BY`).  it looks like there is no validation of the actual qvalues inside the HAVING expression: `SELECT 3 FROM kv GROUP BY k HAVING v > 2` spits out two rows of `3`, as nonsensical as that is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/parser/datum.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jtj6eLRGdEuv78D8i:-K6jtj6eLRGdEuv78D8j:1597374490) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/parser/datum.go#L89)):*
Looks like this is happening independently in #3544 so I'll just wait for that to land.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/group.go, line 54 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jxU-buJGnf3tpAUff:-K6jxU-buJGnf3tpAUfg:1218592726) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f8c48a0b2a29190eec6485aa880e4b0eeda4c77/sql/group.go#L54)):*
#3551 will change this too.

---

*[sql/group.go, line 452 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jxZYg0VIz16fEj151:-K6jxZYhVkXu-wV4vN3O:-1904173360) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f8c48a0b2a29190eec6485aa880e4b0eeda4c77/sql/group.go#L452)):*
ditto above about #3551

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/group.go, line 54 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jxU-buJGnf3tpAUff:-K6jxcT5dca1lMg9cyMR:-1881903867) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f8c48a0b2a29190eec6485aa880e4b0eeda4c77/sql/group.go#L54)):*
(as-in, I'll update to do the right then once #3551 lands)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/group.go, line 54 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jxU-buJGnf3tpAUff:-K6nm6TODifMC5iWwrG6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f8c48a0b2a29190eec6485aa880e4b0eeda4c77/sql/group.go#L54)):*
Done.

---

*[sql/group.go, line 452 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jxZYg0VIz16fEj151:-K6nm8vBvT_9uSMNQJZr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f8c48a0b2a29190eec6485aa880e4b0eeda4c77/sql/group.go#L452)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-:-K6oDZsX85-_2knsTw_o:676834233)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6oQ4OIRDazAj4gFUsF:-767332435) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
I dug into this further and discovered this nonsensical query only works (that is, does not error) with `GROUP BY k`, but fails if, eg you swap all `v` and `k` terms. Digging into the docs, I found this:

> When GROUP BY is present, it is not valid for the SELECT list expressions to refer to ungrouped columns except within aggregate functions or if the ungrouped column is functionally dependent on the grouped columns, since there would otherwise be more than one possible value to return for an ungrouped column. A functional dependency exists if the grouped columns (or a subset thereof) are the primary key of the table containing the ungrouped column.

The last sentence appears to be the key: if the GROUP BY includes the primary key, validation of any qvalue from that table is essentially disabled. Tamir’s recommendation was to hold off on adding this until we find a case which suggests we want it -- until then we're slightly stricter than postgres, in that we just require every field to be aggregated or grouped.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/testdata/aggregate, line 145 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6oTQq7aUd8FrW4yZs_:-K6oTQq7aUd8FrW4yZsa:-1377741847) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac60d2c98b971cbcd5b02429f17ae442ab55c0e5/sql/testdata/aggregate#L145)):*
you're missing the article here, or else "query" should be "queries"

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/testdata/aggregate, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6jJhNV6Pe-wJLlU74s:-K6oVCU7pn5PbmZ7qdKm:1887419520) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc152637666a14a2a903cb4b8c397d10afc4eab/sql/testdata/aggregate#L126)):*
I went ahead and added the validation of HAVING expression, and changed the tests to expect the above example to error.

---

*[sql/testdata/aggregate, line 145 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6oTQq7aUd8FrW4yZs_:-K6oVAPYCx_eAZyV7b1Y:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac60d2c98b971cbcd5b02429f17ae442ab55c0e5/sql/testdata/aggregate#L145)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/group.go, line 36 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547#-K6oVeomwmpN4T6RNa4f:-K6oVeomwmpN4T6RNa4g:602743788) ([raw file](https://github.com/cockroachdb/cockroach/blob/18b291497cc469becbbb8f2d9fe275fa162f2f52/sql/group.go#L36)):*
what's with this space?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3547)*
<!-- Sent from Reviewable.io -->
